--- a/tests/TESTLIST
+++ b/tests/TESTLIST
@@ -36,24 +36,24 @@
 bgp-rt-prefix	bgp-rt-prefix.pcap	bgp-rt-prefix.out	-v
 
 # Broadcom tag tests
-brcmtag		brcm-tag.pcap		brcm-tag.out
-brcmtag-e	brcm-tag.pcap		brcm-tag-e.out	-e
-brcmtagprepend	brcm-tag-prepend.pcap	brcm-tag-prepend.out -e
+#brcmtag		brcm-tag.pcap		brcm-tag.out
+#brcmtag-e	brcm-tag.pcap		brcm-tag-e.out	-e
+#brcmtagprepend	brcm-tag-prepend.pcap	brcm-tag-prepend.out -e
 
 # Broadcom LI
 bcm-li bcm-li.pcap bcm-li.out
 bcm-li-v bcm-li.pcap bcm-li-v.out -v
 
 # Marvell DSA tag tests
-dsa		dsa.pcap		dsa.out
-dsa-e		dsa.pcap		dsa-e.out	-e
+#dsa		dsa.pcap		dsa.out
+#dsa-e		dsa.pcap		dsa-e.out	-e
 
 # EAP tests
 # now in smb.tests
 
 # Marvell DSA tag tests
-edsa		edsa.pcap		edsa.out
-edsa-e		edsa.pcap		edsa-e.out	-e
+#edsa		edsa.pcap		edsa.out
+#edsa-e		edsa.pcap		edsa-e.out	-e
 
 # ESP tests
 esp0		02-sunrise-sunset-esp.pcap	esp0.out
