--- a/tcpdump.c
+++ b/tcpdump.c
@@ -211,6 +211,8 @@
  */
 extern int dflag;
 int dflag;				/* print filter code */
+static int gflag;			/* exit after this many seconds */
+static time_t gflag_time;		/* time_t to exit at */
 static int Gflag;			/* rotate dump files after this many seconds */
 static int Gflag_count;			/* number of files created with Gflag rotation */
 static time_t Gflag_time;		/* The last time_t the dump file was rotated. */
@@ -278,8 +280,10 @@
 #ifdef _WIN32
     static HANDLE timer_handle = INVALID_HANDLE_VALUE;
     static void CALLBACK verbose_stats_dump(PVOID param, BOOLEAN timer_fired);
+    static void CALLBACK gflag_timer_fired(PVOID param, BOOLEAN timer_fired);
 #else /* _WIN32 */
   static void verbose_stats_dump(int sig);
+  static void gflag_timer_fired(int sig);
 #endif /* _WIN32 */
 
 static void info(int);
@@ -672,7 +676,7 @@
 #define U_FLAG
 #endif
 
-#define SHORTOPTS "aAb" B_FLAG "c:C:d" D_FLAG "eE:fF:G:hHi:" I_FLAG j_FLAG J_FLAG "KlLm:M:nNOpq" Q_FLAG "r:s:StT:u" U_FLAG "vV:w:W:xXy:Yz:Z:#"
+#define SHORTOPTS "aAb" B_FLAG "c:C:d" D_FLAG "eE:fF:g:G:hHi:" I_FLAG j_FLAG J_FLAG "KlLm:M:nNOpq" Q_FLAG "r:s:StT:u" U_FLAG "vV:w:W:xXy:Yz:Z:#"
 
 /*
  * Long options.
@@ -1629,6 +1633,20 @@
 
 		case 'F':
 			infile = optarg;
+			break;
+
+		case 'g':
+			gflag = atoi(optarg);
+			if (gflag < 1)
+				error("invalid number of seconds %s", optarg);
+
+			/* Grab the current time for time limit. */
+			if ((gflag_time = time(NULL)) == (time_t)-1) {
+				error("%s: can't get current time: %s",
+				    __func__, pcap_strerror(errno));
+			}
+
+			gflag_time += gflag;
 			break;
 
 		case 'G':
@@ -1636,8 +1654,8 @@
 			if (Gflag < 0)
 				error("invalid number of seconds %s", optarg);
 
-                        /* We will create one file initially. */
-                        Gflag_count = 0;
+			/* We will create one file initially. */
+			Gflag_count = 0;
 
 			/* Grab the current time for rotation use. */
 			if ((Gflag_time = time(NULL)) == (time_t)-1) {
@@ -2509,6 +2527,16 @@
 		timer.it_value.tv_usec = 1;
 		setitimer(ITIMER_REAL, &timer, NULL);
 #endif /* _WIN32 */
+	} else if (gflag > 0) {
+#ifdef _WIN32
+		CreateTimerQueueTimer(&timer_handle, NULL,
+		    gflag_timer_fired, NULL, gflag * 1000, 0,
+		    WT_EXECUTEDEFAULT|WT_EXECUTELONGFUNCTION);
+		setvbuf(stderr, NULL, _IONBF, 0);
+#else /* _WIN32 */
+		(void)setsignal(SIGALRM, gflag_timer_fired);
+		alarm(gflag);
+#endif /* _WIN32 */
 	}
 
 	if (RFileName == NULL) {
@@ -3132,17 +3160,39 @@
 		pcap_dump_flush(pdd);
 }
 #endif
+
+static void
+gflag_time_exceeded(void)
+{
+	(void)fprintf(stderr, "Time limit reached: %d seconds\n", gflag);
+	info(1);
+	exit_tcpdump(0);
+}
 
 static void
 print_packets_captured (void)
 {
 	static u_int prev_packets_captured, first = 1;
+	time_t t;
 
 	if (infodelay == 0 && (first || packets_captured != prev_packets_captured)) {
 		fprintf(stderr, "Got %u\r", packets_captured);
 		first = 0;
 		prev_packets_captured = packets_captured;
 	}
+
+	if (gflag > 0) {
+		/* Get the current time */
+		if ((t = time(NULL)) == (time_t)-1) {
+			error("%s: can't get current_time: %s",
+				__func__, pcap_strerror(errno));
+		}
+
+		/* If the time hits the expected end time, exit */
+		if (t >= gflag_time) {
+			gflag_time_exceeded();
+		}
+	}
 }
 
 /*
@@ -3161,6 +3211,22 @@
 }
 #endif /* _WIN32 */
 
+/*
+ * Called if gflag is set
+ */
+#ifdef _WIN32
+static void CALLBACK gflag_timer_fired(PVOID param _U_,
+    BOOLEAN timer_fired _U_)
+{
+	gflag_time_exceeded();
+}
+#else /* _WIN32 */
+static void gflag_timer_fired(int sig _U_)
+{
+	gflag_time_exceeded();
+}
+#endif /* _WIN32 */
+
 DIAG_OFF_DEPRECATION
 static void
 print_version(FILE *f)
